File: Palindrome Partitioning/Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        allPartitions = []
        curPartition = []

        def isPalindrome(s: str) -> bool:
            i,j = 0, len(s)-1
            
            # check if invalid
            while i <= j:
                if s[i] != s[j]: 
                    return False
                else:
                    i+=1
                    j-=1

            return True # all checks passed

        def dfs(i: int) -> None:
            if i >= len(s):
                allPartitions.append( curPartition[:] )

            # recurse on all palindromes in s[i:]
            for j in range(i, len(s)):
                substr = s[i:j+1]
                if isPalindrome(substr):
                    curPartition.append(substr)
                    dfs( i + len(substr) )
                    curPartition.pop()

        dfs(0)
        return allPartitions

File: Palindrome Partitioning/Palindrome Partitioning.ts
function partition(s: string): string[][] {
    const allPartitions: string[][] = [];
    const curPartition: string[] = [];
    dfs(0);
    return allPartitions;

    function dfs(i: number): void {
        // base case
        if (i === s.length) {
            allPartitions.push([...curPartition]);
            return;
        }

        // ops
        const palindromes: string[] = [];
        for (let j=i; j<s.length; j++) {
            const substr: string = s.slice(i, j+1);
            if (isPalindrome(substr)) palindromes.push(substr);
        }

        // recursion
        for (const palindrome of palindromes) {
            curPartition.push(palindrome);
            dfs(i+palindrome.length);
            curPartition.pop();
        }
    }
};

// helper - palindrome checker
function isPalindrome(s: string): boolean {
    let i: number = 0
    let j: number = s.length-1;

    // find invalidity
    while (i <= j) {
        if (s[i] !== s[j]) return false;
        i++;
        j--;
    }

    // passed all validity checks
    return true;
}